fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/domain/AbstractAuditableWithUTCDateTimeCustom.java

```diff
@@ -0,0 +1,108 @@
+import static org.apache.fineract.infrastructure.core.domain.AuditableFieldsConstants.LAST_MODIFIED_DATE_DB_FIELD;
+
+import java.time.OffsetDateTime;
+import java.util.Optional;
+import javax.persistence.Column;
+import javax.persistence.MappedSuperclass;
+import org.apache.fineract.infrastructure.core.service.DateUtils;
+import org.springframework.data.domain.Auditable;
+import org.springframework.data.jpa.domain.AbstractAuditable;
+
+/**
+ * A custom copy of {@link AbstractAuditable} to override the column names used on database. It also uses OffsetDateTime
+ * for created and modified. The datetimes will be converted from tenant TZ to UTC before storing (automatically
+ * happens) and converted from System TZ to tenant TZ after fetching from DB
+ *
+ * Abstract base class for auditable entities. Stores the audit values in persistent fields.
+ */
+@MappedSuperclass
+public abstract class AbstractAuditableWithUTCDateTimeCustom extends AbstractPersistableCustom
+        implements Auditable<Long, Long, OffsetDateTime> {
+
+    private static final long serialVersionUID = 141481953116476081L;
+
+    @Column(name = CREATED_BY_DB_FIELD, nullable = false)
+    private Long createdBy;
+
+    @Column(name = CREATED_DATE_DB_FIELD, nullable = false)
+    private OffsetDateTime createdDate;
+
+    @Column(name = LAST_MODIFIED_BY_DB_FIELD, nullable = false)
+    private Long lastModifiedBy;
+
+    @Column(name = LAST_MODIFIED_DATE_DB_FIELD, nullable = false)
+    private OffsetDateTime lastModifiedDate;
+
+    @Override
+    public Optional<Long> getCreatedBy() {
+        return Optional.ofNullable(this.createdBy);
+    }
+
+    @Override
+    public void setCreatedBy(final Long createdBy) {
+        this.createdBy = createdBy;
+    }
+
+    @Override
+    public Optional<OffsetDateTime> getCreatedDate() {
+        if (this.createdDate == null) {
+            return Optional.empty();
+        } else {
+            return Optional.of(this.createdDate
+                    .withOffsetSameInstant(DateUtils.getDateTimeZoneOfTenant().getRules().getOffset(this.createdDate.toInstant())));
+        }
+    }
+
+    @Override
+    public void setCreatedDate(final OffsetDateTime createdDate) {
+        this.createdDate = createdDate;
+    }
+
+    @Override
+    public Optional<Long> getLastModifiedBy() {
+        return Optional.ofNullable(this.lastModifiedBy);
+    }
+
+    @Override
+    public void setLastModifiedBy(final Long lastModifiedBy) {
+        this.lastModifiedBy = lastModifiedBy;
+    }
+
+    @Override
+    public Optional<OffsetDateTime> getLastModifiedDate() {
+        if (this.lastModifiedDate == null) {
+            return Optional.empty();
+        } else {
+            return Optional.of(this.lastModifiedDate
+                    .withOffsetSameInstant(DateUtils.getDateTimeZoneOfTenant().getRules().getOffset(this.lastModifiedDate.toInstant())));
```

Comment:
Which by the way is not happening automatically (i mean the UTC), because the OffsetDateTime will use the system TZ when the date time got fetched from the DB...

Reply:
2022-07-26 13:24:12 +5:30 (tenant TZ), the date to be stored
2022-07-26 7:54:12  got stored in DB
2022-07-26 9:54:12 +2:00 (system TZ) will be the OffsetDateTime (after fetched by JPA)